<a href="https://colab.research.google.com/github/qazjean/practicum/blob/main/Lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
print("Первые 7 строк таблицы:")
print(f"{df.head(7)} \n")
print("Техническая сводка о данных")
print(f"{df.info()} \n")
print("Статистическая сводка")
print(df.describe())


Первые 7 строк таблицы:
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         

In [3]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
#задание 2.1
#median_age = df['Age'].median()
#df['Age'].fillna(median_age, inplace=True)

#задание 2.3
df['Age'] = df.groupby(['Sex', 'Pclass'])['Age'].transform(
lambda x: x.fillna(x.median())
)
print(f"{df.info()} \n")

#задание 2.2
def get_age_group(ag):
  if int(ag) < 18:
    return "Ребенок"
  elif int(ag) >= 18 and int(ag) <= 65:
    return "Взрослый"
  else:
    return "Пожилой"
df['AgeGroup'] = df['Age'].apply(get_age_group)
print("\nПервые 5 строк с новым столбцом AgeGroup:")
print(df[['Age', 'AgeGroup']].head())

survival_by_age_group = df.groupby(['AgeGroup'])['Survived'].mean()*100
print(f'\n Процент выживших в зависимости от возрастной группы: \n {survival_by_age_group} \n')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None 


Первые 5 строк с новым столбцом AgeGroup:
    Age  AgeGroup
0  22.0  Взрослый
1  38.0  Взрослый
2  26.0  Взрослый
3  35.0  Взрослый
4  35.0  Взрослый

 Процент выживших в зависимости от возрастной группы: 
 AgeGroup
Взрослый    36.36363

In [8]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
#Задание 3.1
survival_by_sex = df.groupby(['Sex'])['Survived'].mean()*100
print(f'Процент выживших в зависимости от гендера: \n {survival_by_sex} \n')
survival_by_class = df.groupby(['Pclass'])['Survived'].mean()*100
print(f'Процент выживших в зависимости от класса: \n {survival_by_class}')
pivot_survival = df.pivot_table(
values='Survived',
index='Sex',
columns='Pclass',
aggfunc='mean'
) * 100
print(f"\nПроцент выживших по полу и классу:\n{pivot_survival} \n")

#Задание 3.2
kids_3_survived = df[
(df['Age'] < 18) &
(df['Pclass'] == 3) &
(df['Survived'] == 1)
][['Name', 'Age', 'Pclass', 'Survived']].sort_values('Age', ascending=False)
print(f"Выжившие несовершеннолетние в 3 классе: \n {kids_3_survived} \n")

oldest_ne_survived_man = df[(df['Sex'] == 'male') &
(df['Survived'] == 0)].nlargest(1, 'Age')[['Name', 'Age', 'Pclass', 'Survived']]
print(f"Самый пожилой не выживший мужчина: \n{oldest_ne_survived_man} \n")

#Задание 3.3
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

df['IsAlone'] = (df['FamilySize'] == 1).astype(int)
print(f"{df.head(7)} \n")
survival_by_alone = df.groupby('IsAlone')['Survived'].mean() * 100
print(f"\nПроцент выживших в зависимости от одиночества:\n{survival_by_alone}\n")

survival_by_family_size = df.groupby('FamilySize')['Survived'].mean() * 100
print(f"\nПроцент выживших в зависимости от размера семьи:\n{survival_by_family_size}")

Процент выживших в зависимости от гендера: 
 Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64 

Процент выживших в зависимости от класса: 
 Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64

Процент выживших по полу и классу:
Pclass          1          2          3
Sex                                    
female  96.808511  92.105263  50.000000
male    36.885246  15.740741  13.544669 

Выжившие несовершеннолетние в 3 классе: 
                                                 Name    Age  Pclass  Survived
68                   Andersson, Miss. Erna Alexandra  17.00       3         1
208                        Carr, Miss. Helen "Ellen"  16.00       3         1
220                   Sunderland, Mr. Victor Francis  16.00       3         1
156                 Gilnagh, Miss. Katherine "Katie"  16.00       3         1
22                       McGowan, Miss. Anna "Annie"  15.00       3         1
875                 Najib, Miss. Adele Kiam

Да, Путешествующие в одиночку имели другие шансы на выживание, чем семейные пассажиры. У пассажиров с семьей процент выживаемости - 50.564972, а у одиноких - 30.353818.

Часть 4.
1. И пол, и класс каюты оказывали влияние на выживаемость. Пол, однако, в большей степени: 74.203822 процента выживаемость у женщин против 18.890815 процентов выживаемости у мужчин. Различия в зависимости от класса каюты тоже существенны, но в меньшей степени: у первого класса 62.962963 процента, у второго - 47.282609, у третьего - 24.236253.

2. Да, такое утверждение верно. Выживаемость среди детей явно больше:

Взрослый    36.363636

Пожилой     12.500000

Ребенок     53.982301

Выживаемость среди женщин тоже явно больше:

female    74.203822

male      18.890815

Так что да, женщины и дети в первую очередь.